### PR TITLE
[WIP ] System.Linq optimisations via architecture change

### PR DESCRIPTION
**Pull request by manofstick**
(update: sorry if people misinterpreted my message here. This is a radical overhaul of the internal implementation of the System.Linq namespace. It is not complete yet, although I was hoping to spark interest and if someone was excited by the possibilities then maybe get some help finishing this prototype. I am under no illusions that such a radical change has only a small possibly of ultimately being merged, but feel that it may be possible. I feel that this is possible because the API guidelines for the rest of .net explicitly and recommend *against* use of Linq because even in its present state it is too garbage heavy - even when effort has been put in to minimise. So if we choose to just accept that Linq is garbage heavy and then look for alternative solutions then... Well, you have this...)

This is a replacement architecture for `System.Linq`.

This is a proposal/proof-of-concept - think of it as a Christmas gift - hack-upon it during the holidays to see what you think!

This is not complete - but as it is "plug compatible" existing implementations of `IEnumerable`s work seemlessly (although those are areas where performance may not be optimial - note also that I have gutted all the previous performance enhancing code).

This relies *heavily* on generics and has thrown the thought of minimizing garbage out (i.e. things like the initial `IEnumerator` being the same object as the `IEnumerable`).

It is slower on smaller collections (up to maybe a few 10s of items? This is mainly due to the lookups in `JIT_VirtualFunctionPointer` due to the generics use) but faster on larger collections (where the JIT_VirtualFunctionPointer cost drops away). Where the existing System.Linq has a specialized optimization it is pretty hard to beat though. ChainLinq currently doesn't optimize the ElementAt, Last, etc. familiy of optimizations.

So there are plenty of reasons to be skeptical, but maybe, just maybe, it could see the light of day on day!

OK, some details.

So why is existing Linq performance bad and what are the current performance work around? This really boils down to two key ideas. The first is that in abstracting away from the underlying collection means that you loose the benefits benefits of how to access that collection, and secondly, the access through `MoveNext` and `Current` is quite chatty.

`ChainLinq` isolates the collection (which it refers to as a `Consumable`) from the filter/transform  (referred to as a `Link`). `Consumable` is defined as:

    internal abstract class Consumable<T> : IEnumerable<T>
    {
        public abstract Result Consume<Result>(Consumer<T, Result> consumer);

        public abstract IEnumerator<T> GetEnumerator();
        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }

Which you can see is just an `IEnumerable` with an extra function `Consume`. The `Consumer` is an actually such as `Aggregate` or `ToList`. 

Creating Consumable uses the following classes:

    internal abstract class ConsumableForAddition<T> : Consumable<T>
    {
        public abstract Consumable<U> AddTail<U>(ILink<T, U> transform);
    }

    abstract class ConsumableForMerging<T> : ConsumableForAddition<T>
    {
        public abstract object TailLink { get; }
        public abstract Consumable<V> ReplaceTailLink<Unknown, V>(ILink<Unknown, V> newLink);
    }

Where a `ILink<,>` is a function lijke `Select` or `Where`.

Anyway, I have implemented a lot of the Linq functionality using this framework, but I don't think I'll be able to do much over the next while (weeks?) so thought it might be interesting for people to play around with.

But I will try to create some simple performance examples is some follow up posts.

### CODE REVIEW

**Comment by manofstick** [comment]
Taking [LinqOptimizer's `CartLinq`](https://github.com/nessos/LinqOptimizer/blob/master/benchmarks/LinqOptimizer.Benchmarks.CSharp/Program.cs#L102-L108) function (wrapped up a bit differently [here](https://gist.github.com/manofstick/69b489394a172d49f8f6feed3a3ce505))

This is basically:

    (from x in dim1
     from y in dim2
     select x * y).Sum();

Where dim1/dim2 are some arrays (see code)

*64-bit*

size | existing | ChainLinq | %
-|-|-|-
0 | 6107 | 11182 | 183%
10 | 2284 | 1885 | 83%
100 | 3143 | 1382 | 44% 
10000 | 3074 | 1256 | 41%
1000000 | 3033 | 1244 | 41%

*32-bit*

size | existing | ChainLinq | %
-|-|-|-
0 | 6783 | 11206 | 165%
10 | 2070 | 2038 | 98%
100 | 2907 | 1488 | 51%
10000 | 2868 | 1337 | 47%
1000000 | 2821  | 1322 | 47%

**Comment by Joe4evr** [comment]
You should go through the [API Review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) before submitting PRs.

**Comment by manofstick** [comment]
@joe4evr

This is not an API change, it is a performance improvement. What is discussed is internal details.

(Well performance *change* as their are losers - on small enumerations, garbage generated - but this is reliant on the current jit implementation - not suggesting that that changes, but once again it is good for thought)

A potential big change obviously - which may never see the light of day.

File: src/System.Collections.NonGeneric/tests/HashtableTests.cs [hunk 1]
```diff
@@ -1208,16 +1208,16 @@ public void SyncRoot()
             var hash1 = new Hashtable();
             var hash2 = new Hashtable();
 
-            Assert.NotEqual(hash1.SyncRoot, hash2.SyncRoot);
-            Assert.Equal(hash1.SyncRoot.GetType(), typeof(object));
+            Assert.NotSame(hash1.SyncRoot, hash2.SyncRoot);
+            Assert.Equal(hash1.SyncRoot.GetType(), typeof(Hashtable));
```
Comment (Clockwork-Muse) [inline src/System.Collections.NonGeneric/tests/HashtableTests.cs]:
```suggestion
            Assert.IsType<Hashtable>(hash1.SyncRoot);
```
(I have no idea if this actually works here, so be cautions with auto-accepting the suggestion.  You may want `IsAssignableFrom` instead.)

...although this and a few similar changes should really be pulled out into a separate PR.

**Comment by Joe4evr** [comment]
@manofstick 
>This is not an API change, it is a performance improvement.

Irrelevant, as per the API Review Process document:
> Pull requests against **corefx** shouldn't be submitted before getting approval. Also, we don't want to get work in progress (WIP). The reason being that we want to reduce the number pending PRs so that we can focus on the work the community expects we take action on.

**Comment by manofstick** [comment]
@joe4evr

My apologies then.

Is there any other place where one can get the ability to run the whole test suite as per the build server available here other than by creating a PR? (Other than creating my own which is neither in my time envelope in either it's creation or maintenance)?

Ultimately I'm doing this because I believe it have the potential to be beneficial to the .net core project. I am not on anyone's coin to do this - adding friction will just be it's death. 

But if that is the case then such is life.

**Comment by benaadams** [comment]
@dotnet-bot test Linux arm64 Release Build

**Comment by jkoritzinsky** [comment]
@manofstick AFAIK this shouldn't have to go through the API review process if this doesn't change the public API surface. Once it's not a WIP, you should tag the owners of the System.Linq area and then they can take a look.

File: src/System.Collections.NonGeneric/tests/HashtableTests.cs (hunk 1 above)
Comment (manofstick) [reply src/System.Collections.NonGeneric/tests/HashtableTests.cs]: ...totally outside the scope of what this PR is about...

**Comment by manofstick** [comment]
Time for another example, just a "semi-real", `GroupBy()` on a word set, a `Count()` on the elements in each group and finally a `ToDictionary()`. Gist [here](https://gist.github.com/manofstick/86c17a183b000665a13336c1978837ad).

bittage | existing | ChainLinq | %
-|-|-|-
32-bit | 1853 | 1506 | 81%
64-bit | 1630 | 1380  | 85 %

**Comment by manofstick** [comment]
@jamesqo 

This PR might be of interest to you - beware this is still in WIP stage, so whatever disclaimers that entails...

Anyway just as I was hunting for various tests to run, I found your PR #13942.

I cloned the [code from your benchmark](https://github.com/jamesqo/Dotnet/) and put it into a .net core 2.2 app where I ran it against the standard library Linq version and this version.

I just put a Stopwatch around the whole thing to get an aggregate of what was going on. The current library came in around 24s. and with this library it was around 15s (albeit with a lot more garbage, which I'll try and address to some degree if I can going forward.)

Anyway, not sure if this was just an abuse of your benchmark - but thought it might trigger some interest here... (I'm looking for any helpers in finishing this - with the full knowledge that this is an experiment which may never get merged into the main line)

**Comment by manofstick** [comment]
Hi @stephentoub

Not sure who else to hassle, so thought I might as well hassle you... Anyway, this is still very very prototype-y but I'm actually starting to feel reasonably good about it. My original fights with `JIT_VirtualFunctionPointer` have pretty much been resolved (slightly worse from a usage perspective as `TResult Consume<TResult>(Consumer<T, Result> consumer);` has become `void Consume(Consumer<T> consumer);` and the result extracted externally, and some duplication of methods to avoid use when not changing type) but overall small collections are pretty well handled now (although for parity with `Select`/`Where` I'd probably add their own custom `Consumable`s).

I still haven't implement an `OrderBy` `Consumable`, and there are a few other things missing - such as some of the optimizations around `ToArray/ToList` and some others - as well as generally the collection of interfaces for optimizations are more tracer rounds, but its in a state that can be played with.

Anyway, I'm sure you're busy, so I suggest having a boozy Friday lunch so you're not going to get any real work done and have a look :-)

File: eng/dependencies.props [hunk 2]
```diff
@@ -9,8 +9,8 @@
        These ref versions are pulled from https://github.com/dotnet/versions.
   -->
   <PropertyGroup>
-    <CoreFxCurrentRef>5ef2aab13f61e2b43bfff3a6feb2f1964e555845</CoreFxCurrentRef>
-    <CoreClrCurrentRef>5ef2aab13f61e2b43bfff3a6feb2f1964e555845</CoreClrCurrentRef>
+    <CoreFxCurrentRef>22de4f063aa58898fc3f2d489840cd933dcd2d7c</CoreFxCurrentRef>
```
Comment (benaadams) [inline eng/dependencies.props]: Picked up some out of scope changes here (e.g. this file)

File: global.json [hunk 3]
```diff
@@ -5,6 +5,6 @@
   "msbuild-sdks": {
     "Microsoft.DotNet.Arcade.Sdk": "1.0.0-beta.18619.2",
     "Microsoft.DotNet.Helix.Sdk": "1.0.0-beta.18619.2",
-    "Microsoft.NET.Sdk.IL": "3.0.0-preview-27221-02"
+    "Microsoft.NET.Sdk.IL": "3.0.0-preview-27223-02"
```
Comment (benaadams) [inline global.json]: Picked up some out of scope changes here (e.g. this file)

File: src/Common/src/CoreLib/System/Object.cs [hunk 4]
```diff
@@ -22,6 +22,14 @@ public Object()
         {
         }
 
+        // Allow an object to free resources before the object is reclaimed by the GC.
+        // This method's virtual slot number is hardcoded in runtimes. Do not add any virtual methods ahead of this.
+        [NonVersionable]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1821:RemoveEmptyFinalizers")]
+        ~Object()
```
Comment (benaadams) [inline src/Common/src/CoreLib/System/Object.cs]: Picked up some out of scope changes here (e.g. this file)

File: src/Microsoft.IO.Redist/src/Microsoft.IO.Redist.csproj [hunk 5]
```diff
@@ -160,10 +160,10 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.GetVolumeInformation.cs"> 
       <Link>Common\Interop\Windows\Interop.GetVolumeInformation.cs</Link> 
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.FindFirstFileEx.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.FindFirstFileEx.cs">
```
Comment (benaadams) [inline src/Microsoft.IO.Redist/src/Microsoft.IO.Redist.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/Microsoft.Win32.Registry/src/Microsoft.Win32.Registry.csproj [hunk 6]
```diff
@@ -42,46 +42,46 @@
     <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.FormatMessage.cs">
       <Link>Common\CoreLib\Interop\Windows\Interop.FormatMessage.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\advapi32\Interop.RegCloseKey.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Advapi32\Interop.RegCloseKey.cs">
```
Comment (benaadams) [inline src/Microsoft.Win32.Registry/src/Microsoft.Win32.Registry.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/Native/Unix/System.Security.Cryptography.Native/pal_hmac.c [hunk 7]
```diff
@@ -31,7 +31,7 @@ HMAC_CTX* CryptoNative_HmacCreate(const uint8_t* key, int32_t keyLen, const EVP_
 
     if (!ret)
     {
-        free(ctx);
+        HMAC_CTX_free(ctx);
```
Comment (benaadams) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_hmac.c]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Collections.NonGeneric/tests/DictionaryBaseTests.cs [hunk 8]
```diff
@@ -295,8 +295,8 @@ public static void SyncRoot()
             // SyncRoot should be the reference to the underlying dictionary, not to MyDictionary
             var dictBase = new MyDictionary();
             object syncRoot = dictBase.SyncRoot;
-            Assert.NotEqual(syncRoot, dictBase);
-            Assert.Equal(dictBase.SyncRoot, dictBase.SyncRoot);
+            Assert.NotSame(syncRoot, dictBase);
```
Comment (benaadams) [inline src/System.Collections.NonGeneric/tests/DictionaryBaseTests.cs]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Collections.NonGeneric/tests/HashtableTests.cs [hunk 9]
```diff
@@ -1208,16 +1208,16 @@ public void SyncRoot()
             var hash1 = new Hashtable();
             var hash2 = new Hashtable();
 
-            Assert.NotEqual(hash1.SyncRoot, hash2.SyncRoot);
-            Assert.Equal(hash1.SyncRoot.GetType(), typeof(object));
+            Assert.NotSame(hash1.SyncRoot, hash2.SyncRoot);
```
Comment (benaadams) [inline src/System.Collections.NonGeneric/tests/HashtableTests.cs]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Collections.Specialized/tests/StringDictionary/StringDictionary.SyncRootTests.cs [hunk 10]
```diff
@@ -11,13 +11,14 @@ public class StringDictionarySyncRootTests
...
         [InlineData(5)]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)] // Different implementation
         public void SyncRoot(int count)
         {
             StringDictionary stringDictionary1 = Helpers.CreateStringDictionary(count);
             StringDictionary stringDictionary2 = Helpers.CreateStringDictionary(count);
 
             Assert.Same(stringDictionary1.SyncRoot, stringDictionary1.SyncRoot);
-            Assert.IsType<object>(stringDictionary1.SyncRoot);
+            Assert.IsType<Hashtable>(stringDictionary1.SyncRoot);
```
Comment (benaadams) [inline src/System.Collections.Specialized/tests/StringDictionary/StringDictionary.SyncRootTests.cs]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Console/src/System.Console.csproj [hunk 11]
```diff
@@ -66,7 +66,7 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.FillConsoleOutputCharacter.cs">
       <Link>Common\Interop\Windows\Interop.FillConsoleOutputCharacter.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.FileTypes.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.FileTypes.cs">
```
Comment (benaadams) [inline src/System.Console/src/System.Console.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Data.SqlClient/src/System.Data.SqlClient.csproj [hunk 12]
```diff
@@ -218,14 +218,14 @@
   </ItemGroup>
   <!-- SqlFileStream Windows Netcoreapp only -->
   <ItemGroup Condition="'$(TargetsNetCoreApp)' == 'true' AND '$(TargetsWindows)' == 'true'">
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.FileTypes.cs">
-      <Link>Common\CoreLib\Interop\Windows\kernel32\Interop.FileTypes.cs</Link>
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.FileTypes.cs">
```
Comment (benaadams) [inline src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Diagnostics.PerformanceCounter/src/System.Diagnostics.PerformanceCounter.csproj [hunk 13]
```diff
@@ -124,8 +124,8 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.ProcessWaitHandle.cs">
       <Link>Common\Interop\Windows\kernel32\Interop.ProcessWaitHandle.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
-      <Link>Common\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs</Link>
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.Diagnostics.PerformanceCounter/src/System.Diagnostics.PerformanceCounter.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Diagnostics.Process/src/System.Diagnostics.Process.csproj [hunk 14]
```diff
@@ -211,8 +211,8 @@
     <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Interop.BOOL.cs">
       <Link>Common\CoreLib\Interop\Windows\Interop.BOOL.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
-      <Link>Common\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs</Link>
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.Diagnostics.Process/src/System.Diagnostics.Process.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.IO.FileSystem.DriveInfo/src/System.IO.FileSystem.DriveInfo.csproj [hunk 15]
```diff
@@ -41,7 +41,7 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.SetVolumeLabel.cs">
       <Link>Common\Interop\Windows\Interop.SetVolumeLabel.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SetThreadErrorMode.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SetThreadErrorMode.cs">
```
Comment (benaadams) [inline src/System.IO.FileSystem.DriveInfo/src/System.IO.FileSystem.DriveInfo.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.IO.FileSystem.Watcher/src/System.IO.FileSystem.Watcher.csproj [hunk 16]
```diff
@@ -33,7 +33,7 @@
     <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Interop.BOOL.cs">
       <Link>Common\CoreLib\Interop\Windows\Interop.BOOL.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.IO.FileSystem.Watcher/src/System.IO.FileSystem.Watcher.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.IO.FileSystem/src/System.IO.FileSystem.csproj [hunk 17]
```diff
@@ -83,7 +83,7 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.GetVolumeInformation.cs">
       <Link>Common\Interop\Windows\Interop.GetVolumeInformation.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.IO.FileSystem/src/System.IO.FileSystem.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.IO.MemoryMappedFiles/src/System.IO.MemoryMappedFiles.csproj [hunk 18]
```diff
@@ -50,7 +50,7 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.GlobalMemoryStatusEx.cs">
       <Link>Common\Interop\Windows\Interop.GlobalMemoryStatusEx.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.IO.MemoryMappedFiles/src/System.IO.MemoryMappedFiles.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.IO.Pipes/src/System.IO.Pipes.csproj [hunk 19]
```diff
@@ -47,7 +47,7 @@
     <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Interop.BOOL.cs">
       <Link>Common\CoreLib\Interop\Windows\Interop.BOOL.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.IO.Pipes/src/System.IO.Pipes.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.IO.Ports/src/System.IO.Ports.csproj [hunk 20]
```diff
@@ -101,17 +101,17 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.GenericOperations.cs">
       <Link>Common\Interop\Windows\kernel32\Interop.GenericOperations.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.IO.Ports/src/System.IO.Ports.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Net.NameResolution/tests/PalTests/System.Net.NameResolution.Pal.Tests.csproj [hunk 21]
```diff
@@ -126,14 +126,14 @@
     <Compile Include="$(CommonPath)\CoreLib\Microsoft\Win32\SafeHandles\SafeLibraryHandle.cs">
       <Link>Common\CoreLi\Microsoft\Win32\SafeHandles\SafeLibraryHandle.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\Interop\Windows\Kernel32\Interop.GetProcAddress.cs">
-      <Link>Interop\Windows\Kernel32\Interop.GetProcAddress.cs</Link>
+    <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.GetProcAddress.cs">
```
Comment (benaadams) [inline src/System.Net.NameResolution/tests/PalTests/System.Net.NameResolution.Pal.Tests.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Net.NameResolution/src/System.Net.NameResolution.csproj [hunk 22]
```diff
@@ -121,14 +121,14 @@
     <Compile Include="$(CommonPath)\CoreLib\Microsoft\Win32\SafeHandles\SafeLibraryHandle.cs" Condition="'$(TargetGroup)' != 'uap'">
       <Link>Common\CoreLib\Microsoft\Win32\SafeHandles\SafeLibraryHandle.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\Interop\Windows\Kernel32\Interop.GetProcAddress.cs" Condition="'$(TargetGroup)' != 'uap'">
-      <Link>Interop\Windows\Kernel32\Interop.GetProcAddress.cs</Link>
+    <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.GetProcAddress.cs" Condition="'$(TargetGroup)' != 'uap'">
```
Comment (benaadams) [inline src/System.Net.NameResolution/src/System.Net.NameResolution.csproj]: Picked up some out of scope changes here (e.g. this file)

File: src/System.Net.HttpListener/src/System.Net.HttpListener.csproj [hunk 23]
```diff
@@ -192,7 +192,7 @@
     <Compile Include="$(CommonPath)\Interop\Windows\kernel32\Interop.LoadLibraryEx.cs">
       <Link>Common\Interop\Windows\mincore\Interop.LoadLibraryEx.cs</Link>
     </Compile>
-    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\kernel32\Interop.SECURITY_ATTRIBUTES.cs">
+    <Compile Include="$(CommonPath)\CoreLib\Interop\Windows\Kernel32\Interop.SECURITY_ATTRIBUTES.cs">
```
Comment (benaadams) [inline src/System.Net.HttpListener/src/System.Net.HttpListener.csproj]: Picked up some out of scope changes here (e.g. this file)

**Comment by manofstick** [comment]
@benaadams

I have not modified these files. Is it possible that I have merged/rebased/something (or not) poorly?

**Comment by benaadams** [comment]
> I have not modified these files. Is it possible that I have merged/rebased/something (or not) poorly?

Yes you have a couple merge commits in the chain of commits

image

Drop them out and it should be drop out the changes to the files

**Comment by stephentoub** [comment]
@manofstick, thanks for sharing your work.

I've not taken a look yet, but can you help me understand how your approach is fundamentally different from the one LINQ in corefx now uses, with `IIListProvider<T>` and `IPartition<T>`?  Is there something inherently different?  Is it generally the same idea but those interfaces are missing certain calls you'd hope to see optimized?

In general, I do not expect we'd be excited to take a complete re-implementation of LINQ.  We've been ok taking piecemeal optimizations, and if you had some of those you were interested in adding, I'm sure they'd be entertained.  But a complete reimplementation is very likely far out of scope.  It's just too much churn for relatively small performance gains (and some regressions) in an area that's more about productivity than about performance (though of course we like to make it faster whenever we can and the trade-offs are right).

Additionally, the heavy reliance on generics throughout would be a further complication, in particular for AOT builds, where such an explosion of generic instantiations can result in massive size-on-disk increases; we've actually had to disable some of our generics usage in LINQ that was introduced for speed in order to trade it off for size.

cc: @cston, @jaredpar

**Comment by manofstick** [comment]
@stephentoub 

Hi - thanks for asking!

It is, you could argue, inherently different. It is basically a generalization of the specialized optimizations that exist in the current implementation. i.e. you have a `SelectEnumerableIterator`, `SelectArrayIterator`, `SelectListIterator` that tie the `Select` + container. Chain-linq on the other hand ties a container (referred to as `Consumable`) + `chain`, where a chain is any combination of _links_ where a _link_ is any composition of operations - i.e. things like `Select`, `Where, 'Distinct`, etc.. (That said I also have, in order to match performance for the vanilla - ie. non indexed versions - of `Select` and `Where`, specific `Consumables` - I would prefer not to have them, but feel it would be unacceptable to have a performance decrease - for such ubiquitous operators - but they could be dumped if trying to make the smallest-on-disk as possible version...)

A second inherently different feature is that the folding operations - `Aggregate`, `Sum`, `ToList` etc. - are now `push` fed, rather than `pull`ing from an enumerable. This is driven by the `Consume` function of the `Consumable`. `GetEnumerator` is implemented where the Consumables `end-of-the-links` is the `IEnumerator<>` object from which data is pushed to.

Unfortunately, although API compatible, it is not performance compatible, so I do need to fill out whole thing whole rather than their being chunks which could be sliced off as particular improvements - this is why I'm slowing filling the whole things out. I completely understand that as such this might never see the light of day - but hey, maybe it might, or hey it might act to seed other peoples thinking - or just be a fun thing to play around with! (fun?!?!? what are people thinking...)

...and as for performance. Well yes, I'm not (can't!) change the computational complexity of these algorithms, so this of course at best can be a linear factor, but I think it's reasonable... Here using an modest-sized array (1000 element) as the source with some simple cases:

        var r = new Random(42);
        var data = Enumerable.Range(0, 1000).Select(_ => r.NextDouble()).ToArray();

        for (var j = 0; j < 10; ++j)
        {
            var checksum = 0.0;
            var sw = Stopwatch.StartNew();
            for (var idx = 0; idx < 100000; ++idx)
            {
                checksum += data.Sum(); // we'll modify this line...
            }
            Console.WriteLine($"{sw.ElapsedMilliseconds} ({checksum})");
        }

data.XXX.Sum() | chainlinq | linq | %
-|-|-|-
_nothing_ | 327 | 610 | 54%
`Select(x=>x`) | 507 | 753 | 67%
`Select((x,i)=>x)` | 616 | 1444 | 43%
`Where(x=>x>0.5)` | 533 | 799 | 67%
`Where(x=>x>0.5).Take(400)` |  588 | 1237 | 48%
`Where((x,i)=>x>i/1000).Skip(100).Take(400)` | 597 | 1339 | 45%

**Comment by manofstick** [comment]
@stephentoub 

Given my inability to get any single person in the world excited by this, maybe I'll look to spinning it out into it's own nuget package - although keeping this alive with your test servers running is beneficial to its development...

Anyway, in my limited understanding of the MIT License which corefx is licensed under is that this should be OK? i.e. most of the code here is a rewrite to the new underlying architecture, but I would be copying the stubs for the Enumerable extension, some algorithmic stuff from `GroupBy` and `OrderBy` functions as well as the test suite.

So does that fly or not?

**Comment by stephentoub** [comment]
> Anyway, in my limited understanding of the MIT License which corefx is licensed under is that this should be OK?

@richlander, can you comment on this?

**Comment by dzmitry-lahoda** [comment]
@manofstick , so your approach is one by one swap of LINQ so that all tests pass? Or some breaking changes exist?

**Comment by manofstick** [comment]
@dzmitry-lahoda

There should be no breaking changes.

Until it's complete there will be some performance penalty for jumping between the difference implementations - it shouldn't be too bad. There will especially be a penalty as I gutted all the previous optimisations!

In a finished product this should be almost always more performant. One place where it wouldn't be would be where an aggregation function is used without any transforms (no where/select etc) and the underlying enumerable is not something that's optimised for (is not an array, list or output of this library).

**Comment by richlander** [comment]
@manofstick -- You can take parts of corefx with your implementation. No problem. If you do that, you should add a [third party notice](https://github.com/dotnet/corefx/blob/master/THIRD-PARTY-NOTICES.TXT) to your repo. Yours would have a single entry for the corefx repo.

**Comment by danmosemsft** [comment]
@manofstick given the discussion above, would it make sense to close this PR in favor of you creating a Nuget package (which you can certainly link to here)? Then later, when it is more mature and proven, we could possible continue the dicussion?

Thanks again for investing time into this.

**Comment by manofstick** [comment]
Spinning off to seperate project... (when I get time :-)

**Comment by manofstick** [comment]
For anyone interested (crickets!) I have finally started the process of spinning this out to is own project. Few new ideas too... Available [here](https://github.com/manofstick/Cistern.Linq) for anyone who wants to help!

**Pull request closed without merging**